# PROYECTO FINAL - Análisis de datos 

## Setup

Instalacion e importacion de librerias inherentes al proyecto

In [1]:
!pip install pandas==2.2.2

In [2]:
import pandas as pd
import os

## IMPORTO DATASETS ORIGINALES

Cargo los csv que contienen la informacion para realizar el proyecto

### 1: Orders

In [3]:
#Importo datasets como DataFrame y veo las primeras 5 filas
df_orders = pd.read_csv("https://raw.githubusercontent.com/facuescalante/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

### 2: Order items

In [4]:
#Importo datasets como DataFrame y veo las primeras 5 filas
df_order_items = pd.read_csv("https://raw.githubusercontent.com/facuescalante/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

### 3: Products

In [5]:
#Importo datasets como DataFrame y veo las primeras 5 filas
df_products = pd.read_csv("https://raw.githubusercontent.com/facuescalante/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

### 4: Customers

In [6]:
#Importo datasets como DataFrame y veo las primeras 5 filas
df_customers = pd.read_csv("https://raw.githubusercontent.com/facuescalante/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

### 5: Payments

In [7]:
#Importo datasets como DataFrame y veo las primeras 5 filas
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/facuescalante/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

## DATOS del DATAFRAME

Analizo como se componen cada uno de los DataFrame para saber :

* tamaño
* tipo de dato
* cantidad de nulos

### Funcion dataDataFrame

Defino una funcion para reutilizar varias veces, que me indica el tamaño, tipo de dato, cantidad de nulos y si hay duplicados.

In [8]:
def dataDataFrame(df):
    print("--------")
    print("Tamaño del DataFrame")
    print(df.shape)
    print("--------")
    print("--------")
    print("Tipo de datos")
    print(df.info())
    print("--------")
    print("--------")
    print("Cantidad de nulos")
    print(df.isnull().sum())
    print("--------")
    print("--------")
    print("Cantidad de duplicados")
    print(df.duplicated().sum())
    

### Orders

In [9]:
dataDataFrame(df_orders)

--------
Tamaño del DataFrame
(99441, 8)
--------
--------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
--------
--------
Cantidad de nulos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp          

### Order items

In [10]:
dataDataFrame(df_order_items)

--------
Tamaño del DataFrame
(112650, 7)
--------
--------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
--------
--------
Cantidad de nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
--------
--------
Cantidad de duplicados
0


### Products

In [11]:
dataDataFrame(df_products)

--------
Tamaño del DataFrame
(32951, 9)
--------
--------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
--------
--------
Cantidad de nulos
product_id                      0
product_category_name         610
product_name_lenght           6

### Customers

In [12]:
dataDataFrame(df_customers)

--------
Tamaño del DataFrame
(99441, 5)
--------
--------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
--------
--------
Cantidad de nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
--------
--------
Cantidad de duplicados
0


### Payments

In [13]:
dataDataFrame(df_order_payments)

--------
Tamaño del DataFrame
(103886, 5)
--------
--------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
--------
--------
Cantidad de nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
--------
--------
Cantidad de duplicados
0


## TRANSFORMACION de DATAFRAMES

Para que el analisis no tenga incomvenientes a futuro es importante transformar los tipos de datos y asegurar se homogeneidad.

En aquellos DataFrames que haya nulos y la proporcion sea baja, es preferible eliminar dichas filas.

### Funcion transfFecha

Defino una funcion para transformar fechas

In [14]:
def transFecha(df, columna_fechas):
    df[columna_fechas] = df[columna_fechas].apply(pd.to_datetime)
    return df

### Funcion transfString

Defino una funcion para transformar strings

In [15]:
def transfString(df, columna_string):
    for columna in columna_string:
        df[columna] = df[columna].astype("string")
    return df    

### Funcion convertMayuscula

Defino una funcion para convertir a mayusculas

In [16]:
def convertMayuscula(df, columna):
    df[columna] = df[columna].str.upper()
    return df

### Funcion convertMinuscula

Defino una funcion para convertir a minusculas

In [17]:
def convertMinuscula(df, columna):
    df[columna] = df[columna].str.lower()
    return df

### Orders

In [18]:
#Porcentaje de nulos por columna
null_porcent_orders = (df_orders.isnull().sum() / len(df_orders)) * 100

print(null_porcent_orders)

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64


In [19]:
#Filtro las columnas con menos del 10% de nulos
columnas_orders = null_porcent_orders[null_porcent_orders <= 10].index

#Elimino las filas con nulos solo en las columnas seleccionadas
df_orders.dropna(subset=columnas_orders, inplace=True)

#Veo cantidad de nulos
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [20]:
#Listo las columnas que contienen fechas
columnas_fecha_orders = ['order_purchase_timestamp', 'order_approved_at',
                          'order_delivered_carrier_date', 'order_delivered_customer_date',
                          'order_estimated_delivery_date']

#Aplico funcion transfFecha
df_orders = transFecha(df_orders, columnas_fecha_orders)

In [21]:
#Listo las columnas a transformar en string 
columnas_string_orders = ["order_id","customer_id","order_status"]

#Aplico funcion transfString
df_orders = transfString(df_orders, columnas_string_orders)

In [22]:
#Consulto que los tipos de datos sean correctos
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB


In [23]:
#Aplico conversion a minusculas
df_orders = convertMinuscula(df_orders, "order_status")

### Order items

In [24]:
#Listo las columnas que contienen fechas
columnas_fechas_orderItems = ['shipping_limit_date']

#Aplico funcion tranfFecha
df_order_items = transFecha(df_order_items, columnas_fechas_orderItems)

In [25]:
#Listo las columnas a transformar en string 
columnas_string_orderItems = ["order_id","order_item_id","product_id","seller_id"]

#Aplico funcion transfString
df_order_items = transfString(df_order_items, columnas_string_orderItems)

In [26]:
#Consulto que los tipos de datos sean correctos
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  string        
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), string(4)
memory usage: 6.0 MB


### Products

In [27]:
#Porcentaje de nulos en cada columna
null_porcent_products = (df_products.isnull().sum() / len(df_products)) * 100

print(null_porcent_products)

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64


In [28]:
#Filtro las columnas con menos del 10% de nulos
columnas_products = null_porcent_products[null_porcent_products <= 10].index

#Elimino las filas con nulos solo en las columnas seleccionadas
df_products.dropna(subset=columnas_products, inplace=True)

#Veo cantidad de nulos
df_products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [29]:
#Listo las columnas a transformar en string 
columnas_string_products = ["product_id","product_category_name"]

#Aplico funcion transfString
df_products = transfString(df_products, columnas_string_products)

In [30]:
#Consulto que los tipos de datos sean correctos
df_products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB


In [31]:
#Aplico conversion a minusculas
df_products = convertMinuscula(df_products, "product_category_name")

In [32]:
df_products.head(5)

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

### Customers 

In [33]:
#Listo las columnas a transformar en string
columnas_string_customers = ["customer_id","customer_unique_id","customer_zip_code_prefix","customer_city","customer_state"]

#Aplico funcion transfString
df_customers = transfString(df_customers, columnas_string_customers)

In [34]:
#Consulto que los tipos de datos sean correctos
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


In [35]:
#Aplico conversion a minusculas
df_customers = convertMinuscula(df_customers, "customer_city")

In [36]:
#Aplico conversion a mayusculas
df_customers = convertMayuscula(df_customers, "customer_state")

### Payments

In [37]:
#Listo las columnas a transformar en string 
columnas_string_order_payments = ["order_id","payment_type"]

#Aplico funcion transfString
df_order_payments = transfString(df_order_payments, columnas_string_order_payments)

In [38]:
#Aplico conversion a minusculas
df_order_payments = convertMinuscula(df_order_payments, "payment_type")

## ANALISIS de DATAFRAMES

Luego de tranformar los DataFrames hago un análisis de los datos que contienen, para poder extraer información relevante (insights).

### Funcion analisisDataFrame

In [39]:
#Defino una funcion para obtener insights relevantes
def analisisDataFrame(df): #Como parametro uso los dataframes
    print("--------")
    print("Analisis de variables cuantitativas")
    columnas_numericas = df.select_dtypes(include=['int', 'float']).columns
    if not columnas_numericas.empty:
        print(df[columnas_numericas].describe())
    print("--------")
    print("--------")
    print("Analisis de variables cualitativas")
    columnas_string = df.select_dtypes(include=['string']).columns
    if not columnas_string.empty:
        print(df[columnas_string].describe())

### Orders

In [40]:
#Aplico funcion
analisisDataFrame(df_orders)

--------
Analisis de variables cuantitativas
--------
--------
Analisis de variables cualitativas
                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       delivered  
freq          96455  


In [41]:
#Observo el estado de los envios
df_orders["order_status"].value_counts()

order_status
delivered    96455
canceled         6
Name: count, dtype: Int64

Se puede observar que la gran mayoría de los pedidos han sido entregados (delivered) y solo una pequeña cantidad ha sido cancelada (canceled).


In [42]:
#Observo el porcentaje de pedidos entregados
porcentaje_entregado = (df_orders['order_status'] == 'delivered').mean() * 100
porcentaje_entregado

99.9937798695846

La eficiencia del proceso de entrega es optima. 

In [43]:
#observo el porcentaje de pedidos cancelados
porcentaje_cancelado = (df_orders['order_status'] == 'canceled').mean() * 100
porcentaje_cancelado

0.006220130415401043

### Orders items

In [44]:
#Aplico funcion
analisisDataFrame(df_order_items)

--------
Analisis de variables cuantitativas
               price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000
--------
--------
Analisis de variables cualitativas
                                order_id order_item_id  \
count                             112650        112650   
unique                             98666            21   
top     8272b63d03f5f79c56e9e4120aec44ef             1   
freq                                  21         98666   

                              product_id                         seller_id  
count                             112650                            112650  
unique                             32951                              3095  
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b479

El promedio de los precios de los productos es de 120 reales, pero este valor está influenciado por algunos productos que son significativamente más caros. La mayoría de los productos (75%) tienen un precio de 134 reales o menos, y la mediana del precio de los productos es de 75 reales. Esto sugiere que la mayoría de los pedidos se realizan para productos que son relativamente baratos, con solo una pequeña proporción de productos más caros que elevan el promedio.

### Products

In [45]:
#Aplico funcion
analisisDataFrame(df_products)

--------
Analisis de variables cuantitativas
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.63

In [46]:
#Analizo la columna category
df_products["product_category_name"].value_counts()

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: Int64

La categoria mas popular o esencial para los compradores es la "ropa de cama-mesa-bano", seguida de cerca por deporte-ocio, muebles-decoracion y por ultimo belleza y salud. 

### Customers

In [47]:
#Aplico funcion
analisisDataFrame(df_customers)

--------
Analisis de variables cuantitativas
--------
--------
Analisis de variables cualitativas
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746  


In [48]:
#Realizo analisis por regiones de los compradores
df_customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: Int64

SP (São Paulo) es el estado con la mayor cantidad de clientes, con un total de 41,746.
RJ (Rio de Janeiro) y MG (Minas Gerais) siguen a São Paulo con 12,852 y 11,635 clientes respectivamente.
Los estados con menos clientes son AC (Acre), AP (Amapá) y RR (Roraima) con 81, 68 y 46 clientes respectivamente.
Esto sugiere que la mayoría de los clientes se encuentran en los estados de São Paulo, Rio de Janeiro y Minas Gerais. 

### Payments

In [49]:
#Aplico funcion
analisisDataFrame(df_order_payments)

--------
Analisis de variables cuantitativas
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
--------
--------
Analisis de variables cualitativas
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  credit_card
freq                                  29        76795


In [50]:
#Observo tipos de pago
df_order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: Int64

La mayoría de los clientes prefieren pagar con tarjeta de crédito. Los métodos de pago como boleto y vale también son bastante populares, mientras que la tarjeta de débito y los métodos no definidos son menos comunes.

## NUEVOS DATAFRAMES

A partir de los dataset originales y despues de trabajar con las columnas que los componen genero dos nuevos DataFrames que exporto a csv para poder utilizarlos y generar un reporte con visualizaciones.

In [51]:
#Creo la carpeta 'dataset_generados'
if not os.path.exists('dataset_generadosPf'):
    os.makedirs('dataset_generadosPf')

### Orders - Order payments

In [52]:
#Realizo un merge
df_merge_orders_payments = pd.merge(df_orders, df_order_payments, how= "inner", on= "order_id")
df_merge_orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0                    2017-10-18                   1  credit_card   
1                    2017-10-18                   3      voucher   
2                    2017-10-18                   2      voucher   
3                    2018-08-13                   1       boleto   
4                    2018-09-04                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [53]:
#Simplifico la información quitando aquellas columnas que no uso de cara al analisis visual
df_merge_orders_payments.drop(['customer_id','order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'], axis=1, inplace=True) #Indico el nombre de las columnas que voy a mantener los cambios
df_merge_orders_payments.head()

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
1  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
2  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
3  53cdb2fc8bc7dce0b6741e2150273451    delivered      2018-07-24 20:41:37   
4  47770eb9100c2d0c44946d9cf07ec65d    delivered      2018-08-08 08:38:49   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  credit_card                     1          18.12  
1                   3      voucher                     1           2.00  
2                   2      voucher                     1          18.59  
3                   1       boleto                     1         141.46  
4                   1  credit_card                     3         179.12

In [54]:
#Exporto como csv el dataframe
df_merge_orders_payments.to_csv("dataset_generadosPf/orders_payments.csv")

### Order items - Products

In [55]:
#Realizo un merge
df_merge_order_items_products = pd.merge(df_order_items, df_products, how = "inner", on= "product_id")
df_merge_order_items_products.head()

order_id order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214             1   
1  130898c0987d1801452a8ed92a670612             1   
2  532ed5e14e24ae1f0d735b91524b98b9             1   
3  6f8c31653edb8c83e1a739408b5ff750             1   
4  7d19f4ef4d04461989632411b7e588b9             1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
2  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
3  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
4  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   

  shipping_limit_date  price  freight_value product_category_name  \
0 2017-09-19 09:45:35   58.9          13.29            cool_stuff   
1 2017-07-05 02:44:11   55.9          17.96            cool_stuff   
2 2018-05-23 10:56:25   64.9          18.33            cool_stuff   
3 2017-08-07 18:55:08   58.9          16.17            cool_stuff   
4 2017-08-16 22:05:11   58.9          13.29            cool_stuff   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 58.0                       598.0                 4.0   
2                 58.0                       598.0                 4.0   
3                 58.0                       598.0                 4.0   
4                 58.0                       598.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1             650.0               28.0                9.0              14.0  
2             650.0               28.0                9.0              14.0  
3             650.0               28.0                9.0              14.0  
4             650.0               28.0                9.0              14.0

In [56]:
#Simplifico la información quitando aquellas columnas que no uso de cara al analisis visual
df_merge_order_items_products.drop(['product_id','seller_id','shipping_limit_date'], axis=1, inplace=True) #Indico el nombre de las columnas que voy a mantener los cambios
df_merge_order_items_products.head(4)

order_id order_item_id  price  freight_value  \
0  00010242fe8c5a6d1ba2dd792cb16214             1   58.9          13.29   
1  130898c0987d1801452a8ed92a670612             1   55.9          17.96   
2  532ed5e14e24ae1f0d735b91524b98b9             1   64.9          18.33   
3  6f8c31653edb8c83e1a739408b5ff750             1   58.9          16.17   

  product_category_name  product_name_lenght  product_description_lenght  \
0            cool_stuff                 58.0                       598.0   
1            cool_stuff                 58.0                       598.0   
2            cool_stuff                 58.0                       598.0   
3            cool_stuff                 58.0                       598.0   

   product_photos_qty  product_weight_g  product_length_cm  product_height_cm  \
0                 4.0             650.0               28.0                9.0   
1                 4.0             650.0               28.0                9.0   
2                 4.0             650.0               28.0                9.0   
3                 4.0             650.0               28.0                9.0   

   product_width_cm  
0              14.0  
1              14.0  
2              14.0  
3              14.0

In [57]:
#Exporto como csv el dataframe
df_merge_order_items_products.to_csv("dataset_generadosPf/order_items_products.csv")

### Orders - Customers

In [60]:
#Realizo un merge
df_merge_orders_customers = pd.merge(df_orders, df_customers, how = "inner", on= "customer_id")
df_merge_orders_customers.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                customer_unique_id  \
0                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
1                    2018-08-13  af07308b275d755c9edb36a90c618231   
2                    2018-09-04  3a653a41f6f9fc3d2a113cf8398680e8   
3                    2017-12-15  7c142cf63193a1473d2e66489a9ae977   
4                    2018-02-26  72632f0f9dd73dfee390c9b22eb56dd6   

  customer_zip_code_prefix            customer_city customer_state  
0                     3149                sao paulo             SP  
1                    47813                barreiras             BA  
2                    75265               vianopolis             GO  
3                    59296  sao goncalo do amarante             RN  
4                     9195              santo andre             SP

In [61]:
#Exporto como csv el dataframe
df_merge_orders_customers.to_csv("dataset_generadosPf/orders_customers.csv")